### ISSUE TITLE
Multiple Enums Copying Incorrectly to Cluster

### ISSUE THREAD

**Issue by WesleyTangNationalInstruments** [ni/grpc-labview#295]
This bug was found in InstrumentStudio using MeasurementLink LabVIEW where multiple consecutive enums were not being sent from the client (InstrumentStudio) to the server (MeasurementLink LabVIEW measurement) correctly. Debugging into this, it seems that the issue lies somewhere grpc-labview where the only the first enum's values get transcribed into the cluster structure. Sumit Jain (@ni-sujain) asked me to recreate this bug item directly in the grpc-labview repo.

Further details and an example measurement that demonstrates this issue can be found in this [AzDO bug item](https://ni.visualstudio.com/DevCentral/_workitems/edit/2478685). 

For completeness, I've also detailed the issue and debugging steps taken below.

------------------------------------------------------

## Bug Description

This behavior was reported by Andy Tan Wei Keat (andy.wei.keat.tan@ni.com) using MeasurementLink LabVIEW v1.2.0.2 which is built against grpc-labview v1.0.0.1. An example measurement was provided to reproduce this issue and attached. 

[Sample Plugin 1001.zip](https://github.com/ni/grpc-labview/files/12372331/Sample.Plugin.1001.zip)

Note that MeasurementLink LabVIEW is currently incompatible with the latest version of grpc-labview and cannot be upgraded from v1.0.0.1.

This example sets up two enums where the configuration pipes directly into the result.

![673cbc01-fdd5-4277-ae8b-8826d8a47dfb](https://github.com/ni/grpc-labview/assets/103216564/e5819053-43ba-4b0c-b35c-2c68685235f1)

When we have two continuous enums in a Measurement Configuration, the second enum will not receive the data in the Measurement Result.

![3d6a27ee-d394-4647-b6d1-a0d8915b42fc](https://github.com/ni/grpc-labview/assets/103216564/9ca576e1-506e-47f5-a97b-f28b8bd8da16)

Note that if a non-enum control is inserted in between the two enum controls, the data is passed as expected.

![2f23b91d-e424-439f-bfdb-7c05a4bb4a00](https://github.com/ni/grpc-labview/assets/103216564/3010826b-e48a-4f62-9ac0-507063718917)

![92b83e06-8d82-438f-a535-e5dff736a715](https://github.com/ni/grpc-labview/assets/103216564/542b9578-413d-40a0-a3d7-c1d4aa214583)

------------------------------------------------------

## Debugging Steps Taken

On the grpc client side, InstrumentStudio correctly serializes the measurement's configuration parameter values. In this example, there are two bytes per parameter, the first byte for the parameter id and `Google.Protobuf.WireFormat.WireType` (in this case, the wire type is VarInt = 0 so the entire first byte can be considered the parameter id) and the second byte for the value. 

![dc5a9cd9-ffb6-4457-a5f2-5f8ea6717ed9](https://github.com/ni/grpc-labview/assets/103216564/cbe95915-253f-4254-93c0-9266ae40af99)
![32159ae5-468d-45d6-9aa9-69ed4239448d](https://github.com/ni/grpc-labview/assets/103216564/2989dd5e-d73d-4e37-809d-5c4eb11a2d04)

On the grpc server side, MeasurementLink LabVIEW correctly receives the serialized configuration parameter message. It then takes that message and passes it into `grpc-lvsupport-release.lvlib:UnpackFromAny.vim`. This eventually calls down into the `UnpackFromBuffer method` in the `labview_grpc_server.dll` assembly.

![628eb4c9-ea5f-44c0-a352-a449624a4874](https://github.com/ni/grpc-labview/assets/103216564/52524690-edca-42a8-a2f6-7dc1ec29c623)

Here the unpacked value only contains the information for the first enum, but the second enum's value is left as default which is incorrect.

It looks like we are getting the correct message in [UnpackFromBuffer ](https://github.com/ni/grpc-labview/blob/master/src/any_support.cc#L78)and the issue likely isn't in the transmission.

![38fd786d-4061-4672-ad3a-90fea4001ac4](https://github.com/ni/grpc-labview/assets/103216564/84506c02-4a8a-46fd-b17b-6a84b002209a)

Here we found an issue where it looks like we are copying the value into the cluster as a [LVUInt32MessageValue](https://github.com/ni/grpc-labview/blob/master/src/cluster_copier.cc#L377) rather than a [LVEnumMessageValue ](https://github.com/ni/grpc-labview/blob/master/src/cluster_copier.cc#L397)(which has int32 underlying), however it appears to have been fixed in a newer version of grpc-labview  (#263 and #283). Manually applying these changesets to v1.0.0.1 resolves this typing weirdness, but does not affect the described problem.

![d0f7d041-d287-482a-9ed3-487dce98ec3b](https://github.com/ni/grpc-labview/assets/103216564/7204ff69-54d6-4894-a663-a27232aa2b02)

Looking at the data we're copying into the cluster itself, it looks like we're correctly using 4 bytes for int32s as expected. One thing to note is that the widths of these values are sourced from clusterOffset which comes from the metadata of the message rather than calling `size()`, though I don't think this is relevant as the 4-byte widths look correct.

![b6b19c70-5285-434c-abd3-9eeee3f80141](https://github.com/ni/grpc-labview/assets/103216564/08a6aee0-4d88-4b60-b584-5a7d76cea5c3)

The missing data for the second enum appears after the values are copied into the cluster. Maybe this is because the cluster is a `int8_t` or because the debugger simply interprets it as a pointer to a value rather than a pointer to an array (or cluster structure, I'm not really sure how this works here). 

![8fc927d8-a0d9-45d2-a5ea-af68ec6c26fa](https://github.com/ni/grpc-labview/assets/103216564/bbb756a1-2137-40f4-84c0-e3276813aaa6)

AB#2490682

**Comment by pratheekshasn** [comment]
I tried reproducing this on my machine with a simple proto file that has two enums next to each inside the same message. I didn't not face any issues sending across enum values from the server to the client, even though there was no numeric field in between the enums.

Could you please attach the proto file that you are using for the project that you have attached?

I will try to install Measurement Link on my machine and debug the attached project, but in the meantime, it would be helpful if I could see the proto file as well.

**Comment by pratheekshasn** [comment]
# TL;DR
## Solution:
In the typedefs, Measurement Configuration.ctl and Measure Results.ctl, change the representation of the enum  from U16 to U32.
## Reason:
This is happening because of mismatch in the numeric representation of the enum type across the Measurement UI code and the gRPC LabVIEW code.

________________________________________________________________________________________________________________________________________________________________________

# Details
gRPC LabVIEW code assumes that the enum types passed to it are 32-bit integers. This is because proto files define enums as 32-bit integers.

The enums in the Measurement UI side of the code, however were 16-bit integers. These are the enums present in the typedefs, Measurement Configuration.ctl and Measure Results.ctl.

When the client is run, the protobuf is parsed by gRPC LabVIEW and the value is copied into a cluster on the Measurement UI code (MeasurementPlugin.lvclass::Unpack For Grpc.vi). This VI internally calls UnpackFromAny.vim, which actually parses the protobuf and returns the value as a LabVIEW cluster. The API in labview_grpc_server.dll writes the correct value of the protobuf into the cluster pointer. But once the cluster is returned to the LabVIEW side, and typecast to one of the typdefs mentioned above, it would read only the first [16 + 16] bytes, as opposed to the first [32 + 32] bytes of actual value. This was why the second enum value would read 0 and just select the default value of the enum.

Changing the representation of the enum in the typdef fixes the issue.

**Comment by pratheekshasn** [comment]
Since this is a fix that needs to go into the attached project, and not into the gRPC LabVIEW code, closing this bug.
